flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -0,0 +1,317 @@
+import java.util.stream.IntStream;
+
+import static org.apache.flink.configuration.JobManagerOptions.FORCE_PARTITION_RELEASE_ON_CONSUMPTION;
+import static org.apache.flink.runtime.executiongraph.failover.FailoverStrategyLoader.PIPELINED_REGION_RESTART_STRATEGY_NAME;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+public class BatchFineGrainedRecoveryITCase extends TestLogger {
+	private static final int EMITTED_RECORD_NUMBER = 1000;
+	private static final int MAX_FAILURE_NUMBER = 10;
+	private static final int MAP_NUMBER = 3;
+	private static final String CHECK_ACC_NAME = "CHECK_ACC_NAME";
+	private static final int[] EXPECTED_JOB_OUTPUT = IntStream.range(MAP_NUMBER, EMITTED_RECORD_NUMBER + MAP_NUMBER).toArray();
+
+	private static TestingMiniCluster miniCluster;
+
+	@Before
+	public void setup() throws Exception {
+		Configuration configuration = new Configuration();
+		configuration.setBoolean(FORCE_PARTITION_RELEASE_ON_CONSUMPTION, false);
+		configuration.setString(JobManagerOptions.EXECUTION_FAILOVER_STRATEGY, PIPELINED_REGION_RESTART_STRATEGY_NAME);
+
+		miniCluster = new TestingMiniCluster(
+			new Builder()
+				.setNumTaskManagers(MAP_NUMBER)
+				.setNumSlotsPerTaskManager(1)
+				.setConfiguration(configuration)
+				.build(),
+			null);
+
+		miniCluster.start();
+	}
+
+	@After
+	public void teardown() throws Exception {
+		if (miniCluster != null) {
+			miniCluster.close();
+		}
+	}
+
+	@Test
+	public void testProgram() throws Exception {
+		ExecutionEnvironment env = createExecutionEnvironment();
+
+		StaticFailureCounter.reset();
+		StaticMapFailureTracker.reset();
+
+		FailureStrategy failureStrategy = new RandomExceptionFailureStrategy(1, EMITTED_RECORD_NUMBER);
+
+		DataSet<Integer> input = env.createInput(FixedRecordNumberInputFormat.emitFixedRecordNumber());
+		for (int i = 0; i < MAP_NUMBER; i++) {
+			input = input.flatMap(new TestMapper(StaticMapFailureTracker.addNewMap(), failureStrategy));
+			input = disableChaining(input, Integer.class);
+		}
+		input.output(new AccumulatingRichOutputFormat());
+
+		verifyResult(env.execute());
+
+		StaticMapFailureTracker.verify();
+	}
+
+	private static ExecutionEnvironment createExecutionEnvironment() {
+		@SuppressWarnings("StaticVariableUsedBeforeInitialization")
+		ExecutionEnvironment env = new TestEnvironment(miniCluster, 1, true);
+		env.setRestartStrategy(RestartStrategies.fixedDelayRestart(MAX_FAILURE_NUMBER, Time.milliseconds(10)));
+		env.getConfig().setExecutionMode(ExecutionMode.BATCH_FORCED);
+		return env;
+	}
+
+	private static <T> DataSet<T> disableChaining(
+			DataSet<T> input,
+			@SuppressWarnings("SameParameterValue") Class<T> cl) {
+		return input.mapPartition((MapPartitionFunction<T, T>) (ns, out) -> ns.forEach(out::collect)).returns(cl);
+	}
+
+	private static void verifyResult(JobExecutionResult result) {
+		Collection<Integer> accumulatorResult = result.getAccumulatorResult(CHECK_ACC_NAME);
```

Comment:
you could simply this by using `DataSet#collect`. You wouldn't need a separate outputformat class, but the behavior should be the same.